File: packages/core/src/trial/designs.jl
# Trial Designs
# Convenience functions for creating study designs

export parallel_design, crossover_2x2, crossover_3x3, williams_design
export dose_escalation_3plus3, dose_escalation_mtpi, dose_escalation_crm
export adaptive_design, bioequivalence_design
export get_design_description

"""
    parallel_design(n_arms; kwargs...)

Create a parallel group study design.

# Arguments
- `n_arms::Int`: Number of treatment arms

# Keyword Arguments
- `randomization_ratio::Vector{Float64}`: Randomization ratio (default: equal)
- `stratification_factors::Vector{Symbol}`: Stratification factors

# Returns
- `ParallelDesign`: Parallel study design

# Example
```julia
# 2-arm parallel with 1:1 randomization
design = parallel_design(2)

# 3-arm with 2:1:1 randomization
design = parallel_design(3, randomization_ratio=[0.5, 0.25, 0.25])
```
"""
function parallel_design(n_arms::Int;
        randomization_ratio::Union{Nothing, Vector{Float64}} = nothing,
        stratification_factors::Vector{Symbol} = Symbol[])
    if randomization_ratio === nothing
        randomization_ratio = ones(n_arms) / n_arms
    end
    return ParallelDesign(n_arms;
        randomization_ratio = randomization_ratio,
        stratification_factors = stratification_factors)
end


"""
    crossover_2x2(; washout_duration=7.0)

Create a 2-period, 2-sequence crossover design (AB, BA).

# Keyword Arguments
- `washout_duration::Float64`: Washout period in days (default: 7.0)

# Returns
- `CrossoverDesign`: 2x2 crossover design

# Example
```julia
design = crossover_2x2(washout_duration=14.0)
```
"""
function crossover_2x2(; washout_duration::Float64 = 7.0)
    return CrossoverDesign(2, 2;
        washout_duration = washout_duration,
        sequence_assignments = [[1, 2], [2, 1]])
end


"""
    crossover_3x3(; washout_duration=7.0)

Create a 3-period, 3-sequence crossover design (Latin square).

# Keyword Arguments
- `washout_duration::Float64`: Washout period in days (default: 7.0)

# Returns
- `CrossoverDesign`: 3x3 crossover design

# Example
```julia
design = crossover_3x3(washout_duration=21.0)
```
"""
function crossover_3x3(; washout_duration::Float64 = 7.0)
    return CrossoverDesign(3, 3;
        washout_duration = washout_duration,
        sequence_assignments = [[1, 2, 3], [2, 3, 1], [3, 1, 2]])
end


"""
    williams_design(n_treatments; washout_duration=7.0)

Create a Williams design for crossover studies.

Williams designs are balanced for first-order carryover effects.

# Arguments
- `n_treatments::Int`: Number of treatments (2, 3, or 4)

# Keyword Arguments
- `washout_duration::Float64`: Washout period in days (default: 7.0)

# Returns
- `CrossoverDesign`: Williams crossover design

# Example
```julia
# 4-treatment Williams design
design = williams_design(4, washout_duration=14.0)
```
"""
function williams_design(n_treatments::Int; washout_duration::Float64 = 7.0)
    if n_treatments == 2
        sequences = [[1, 2], [2, 1]]
    elseif n_treatments == 3
        sequences = [[1, 2, 3], [2, 3, 1], [3, 1, 2],
                     [1, 3, 2], [3, 2, 1], [2, 1, 3]]
    elseif n_treatments == 4
        sequences = [[1, 2, 4, 3], [2, 3, 1, 4], [3, 4, 2, 1], [4, 1, 3, 2],
                     [1, 4, 2, 3], [2, 1, 3, 4], [3, 2, 4, 1], [4, 3, 1, 2]]
    else
        error("Williams design only supports 2, 3, or 4 treatments")
    end

    return CrossoverDesign(n_treatments, length(sequences);
        washout_duration = washout_duration,
        sequence_assignments = sequences)
end


"""
    dose_escalation_3plus3(dose_levels; kwargs...)

Create a 3+3 dose escalation design.

# Arguments
- `dose_levels::Vector{Float64}`: Available dose levels

# Keyword Arguments
- `starting_dose::Float64`: Starting dose (default: first level)
- `max_dlt_rate::Float64`: Maximum acceptable DLT rate (default: 0.33)
- `cohort_size::Int`: Cohort size (default: 3)
- `max_subjects::Int`: Maximum total subjects (default: 30)

# Returns
- `DoseEscalationDesign`: 3+3 dose escalation design

# Example
```julia
design = dose_escalation_3plus3([10.0, 25.0, 50.0, 100.0, 200.0])
```
"""
function dose_escalation_3plus3(dose_levels::Vector{Float64};
        starting_dose::Union{Float64, Nothing} = nothing,
        max_dlt_rate::Float64 = 0.33,
        cohort_size::Int = 3,
        max_subjects::Int = 30)
    rule = ThreePlusThree(max_dlt_rate = max_dlt_rate)
    return DoseEscalationDesign(dose_levels;
        starting_dose = starting_dose,
        escalation_rule = rule,
        cohort_size = cohort_size,
        max_subjects = max_subjects)
end


"""
    dose_escalation_mtpi(dose_levels; kwargs...)

Create an mTPI (modified Toxicity Probability Interval) dose escalation design.

# Arguments
- `dose_levels::Vector{Float64}`: Available dose levels

# Keyword Arguments
- `starting_dose::Float64`: Starting dose (default: first level)
- `target_dlt_rate::Float64`: Target DLT rate (default: 0.25)
- `equivalence_interval::Tuple{Float64, Float64}`: Equivalence interval (default: (0.20, 0.30))
- `cohort_size::Int`: Cohort size (default: 3)
- `max_subjects::Int`: Maximum total subjects (default: 30)

# Returns
- `DoseEscalationDesign`: mTPI dose escalation design

# Example
```julia
design = dose_escalation_mtpi([10.0, 25.0, 50.0, 100.0, 200.0],
                               target_dlt_rate=0.30)
```
"""
function dose_escalation_mtpi(dose_levels::Vector{Float64};
        starting_dose::Union{Float64, Nothing} = nothing,
        target_dlt_rate::Float64 = 0.25,
        equivalence_interval::Tuple{Float64, Float64} = (0.20, 0.30),
        cohort_size::Int = 3,
        max_subjects::Int = 30)
    rule = mTPI(target_dlt_rate = target_dlt_rate,
                equivalence_interval = equivalence_interval)
    return DoseEscalationDesign(dose_levels;
        starting_dose = starting_dose,
        escalation_rule = rule,
        cohort_size = cohort_size,
        max_subjects = max_subjects)
end


"""
    dose_escalation_crm(dose_levels; kwargs...)

Create a CRM (Continual Reassessment Method) dose escalation design.

# Arguments
- `dose_levels::Vector{Float64}`: Available dose levels

# Keyword Arguments
- `starting_dose::Float64`: Starting dose (default: first level)
- `target_dlt_rate::Float64`: Target DLT rate (default: 0.25)
- `skeleton::Vector{Float64}`: Prior toxicity probabilities
- `model::Symbol`: Dose-toxicity model (:logistic or :power)
- `cohort_size::Int`: Cohort size (default: 1)
- `max_subjects::Int`: Maximum total subjects (default: 30)

# Returns
- `DoseEscalationDesign`: CRM dose escalation design

# Example
```julia
design = dose_escalation_crm([10.0, 25.0, 50.0, 100.0, 200.0],
                              skeleton=[0.05, 0.12, 0.25, 0.40, 0.55])
```
"""
function dose_escalation_crm(dose_levels::Vector{Float64};
        starting_dose::Union{Float64, Nothing} = nothing,
        target_dlt_rate::Float64 = 0.25,
        skeleton::Union{Nothing, Vector{Float64}} = nothing,
        model::Symbol = :logistic,
        cohort_size::Int = 1,
        max_subjects::Int = 30)
    if skeleton === nothing
        # Generate default skeleton
        n = length(dose_levels)
        skeleton = [0.05 + 0.9 * (i - 1) / (n - 1) * target_dlt_rate for i in 1:n]
    end
    rule = CRM(target_dlt_rate = target_dlt_rate,
               skeleton = skeleton,
               model = model)
    return DoseEscalationDesign(dose_levels;
        starting_dose = starting_dose,
        escalation_rule = rule,
        cohort_size = cohort_size,
        max_subjects = max_subjects)
end


"""
    adaptive_design(base_design; kwargs...)

Create an adaptive trial design.

# Arguments
- `base_design::StudyDesignKind`: Underlying study design

# Keyword Arguments
- `interim_analyses::Vector{Float64}`: Information fractions for interim analyses
- `alpha_spending::Symbol`: Alpha spending function (:obrien_fleming, :pocock, :haybittle_peto)
- `futility_boundary::Float64`: Futility boundary (conditional power threshold)
- `sample_size_reestimation::Bool`: Allow sample size re-estimation

# Returns
- `AdaptiveDesign`: Adaptive trial design

# Example
```julia
base = parallel_design(2)
design = adaptive_design(base, interim_analyses=[0.5],
                         alpha_spending=:obrien_fleming)
```
"""
function adaptive_design(base_design::StudyDesignKind;
        interim_analyses::Vector{Float64} = [0.5],
        alpha_spending::Symbol = :obrien_fleming,
        futility_boundary::Float64 = 0.10,
        sample_size_reestimation::Bool = false)

    adaptation_rules = Dict{Symbol, Any}(
        :futility_boundary => futility_boundary,
        :sample_size_reestimation => sample_size_reestimation
    )

    return AdaptiveDesign(base_design;
        interim_analyses = interim_analyses,
        adaptation_rules = adaptation_rules,
        alpha_spending = alpha_spending)
end


"""
    bioequivalence_design(; kwargs...)

Create a bioequivalence study design.

# Keyword Arguments
- `n_periods::Int`: Number of periods (default: 2)
- `n_sequences::Int`: Number of sequences (default: 2)
- `washout_duration::Float64`: Washout period in days (default: 7.0)
- `bioequivalence_limits::Tuple{Float64, Float64}`: BE limits (default: (0.80, 1.25))
- `parameters::Vector{Symbol}`: PK parameters to assess (default: [:cmax, :auc_0_inf])
- `regulatory_guidance::Symbol`: Regulatory guidance (:fda, :ema)

# Returns
- `BioequivalenceDesign`: Bioequivalence study design

# Example
```julia
# FDA BE study
design = bioequivalence_design(regulatory_guidance=:fda)

# EMA BE study with highly variable drug widened limits
design = bioequivalence_design(
    bioequivalence_limits=(0.6984, 1.4319),
    regulatory_guidance=:ema
)
```
"""
function bioequivalence_design(;
        n_periods::Int = 2,
        n_sequences::Int = 2,
        washout_duration::Float64 = 7.0,
        bioequivalence_limits::Tuple{Float64, Float64} = (0.80, 1.25),
        parameters::Vector{Symbol} = [:cmax, :auc_0_inf],
        regulatory_guidance::Symbol = :fda)

    return BioequivalenceDesign(;
        n_periods = n_periods,
        n_sequences = n_sequences,
        washout_duration = washout_duration,
        bioequivalence_limits = bioequivalence_limits,
        parameters = parameters,
        regulatory_guidance = regulatory_guidance)
end


"""
    get_design_description(design::StudyDesignKind)

Get a human-readable description of a study design.

# Arguments
- `design::StudyDesignKind`: Study design

# Returns
- `String`: Design description
"""
function get_design_description(design::ParallelDesign)
    return "$(design.n_arms)-arm parallel group design"
end

function get_design_description(design::CrossoverDesign)
    return "$(design.n_periods)-period $(design.n_sequences)-sequence crossover design"
end

function get_design_description(design::DoseEscalationDesign)
    rule_name = typeof(design.escalation_rule) |> string |> x -> replace(x, "OpenPKPDCore." => "")
    return "Dose escalation ($(rule_name)) with $(length(design.dose_levels)) dose levels"
end

function get_design_description(design::AdaptiveDesign)
    base_desc = get_design_description(design.base_design)
    n_interim = length(design.interim_analyses)
    return "Adaptive $(base_desc) with $(n_interim) interim analysis(es)"
end

function get_design_description(design::BioequivalenceDesign)
    return "Bioequivalence $(get_design_description(design.crossover)) ($(design.regulatory_guidance))"
end

File: packages/python/openpkpd/trial/designs.py
"""
OpenPKPD Trial Designs

Study design definitions for clinical trials including:
- Parallel group designs
- Crossover designs (2x2, 3x3, Williams)
- Dose escalation (3+3, mTPI, CRM)
- Adaptive designs
- Bioequivalence designs
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any, Dict, List, Optional, Tuple, Union
from enum import Enum

class EscalationRuleType(str, Enum):
    """Dose escalation rule types."""
    THREE_PLUS_THREE = "3+3"
    MTPI = "mTPI"
    CRM = "CRM"


@dataclass
class ParallelDesign:
    """
    Parallel group study design.

    Attributes:
        n_arms: Number of treatment arms
        randomization_ratio: Randomization ratio for each arm
        stratification_factors: Factors for stratified randomization

    Example:
        >>> design = ParallelDesign(n_arms=2)
        >>> design = ParallelDesign(n_arms=3, randomization_ratio=[0.5, 0.25, 0.25])
    """
    n_arms: int
    randomization_ratio: List[float] = field(default_factory=list)
    stratification_factors: List[str] = field(default_factory=list)

    def __post_init__(self):
        if not self.randomization_ratio:
            self.randomization_ratio = [1.0 / self.n_arms] * self.n_arms


@dataclass
class CrossoverDesign:
    """
    Crossover study design.

    Attributes:
        n_periods: Number of treatment periods
        n_sequences: Number of treatment sequences
        washout_duration: Washout period duration in days
        sequence_assignments: Treatment assignments per sequence

    Example:
        >>> design = CrossoverDesign(n_periods=2, n_sequences=2)  # 2x2 crossover
    """
    n_periods: int
    n_sequences: int
    washout_duration: float = 7.0
    sequence_assignments: List[List[int]] = field(default_factory=list)


@dataclass
class DoseEscalationDesign:
    """
    Dose escalation study design for Phase I trials.

    Attributes:
        dose_levels: Available dose levels
        starting_dose: Starting dose level
        escalation_rule: Rule type ('3+3', 'mTPI', 'CRM')
        cohort_size: Number of subjects per cohort
        max_subjects: Maximum total subjects
        target_dlt_rate: Target DLT rate (for mTPI/CRM)

    Example:
        >>> design = DoseEscalationDesign(
        ...     dose_levels=[10.0, 25.0, 50.0, 100.0, 200.0],
        ...     escalation_rule='3+3'
        ... )
    """
    dose_levels: List[float]
    starting_dose: Optional[float] = None
    escalation_rule: str = "3+3"
    cohort_size: int = 3
    max_subjects: int = 30
    target_dlt_rate: float = 0.25

    def __post_init__(self):
        if self.starting_dose is None:
            self.starting_dose = self.dose_levels[0]


@dataclass
class AdaptiveDesign:
    """
    Adaptive trial design with interim analyses.

    Attributes:
        base_design: Underlying study design
        interim_analyses: Information fractions for interim analyses
        alpha_spending: Alpha spending function type
        futility_boundary: Futility boundary threshold

    Example:
        >>> base = ParallelDesign(n_arms=2)
        >>> design = AdaptiveDesign(base_design=base, interim_analyses=[0.5])
    """
    base_design: Union[ParallelDesign, CrossoverDesign]
    interim_analyses: List[float] = field(default_factory=lambda: [0.5])
    alpha_spending: str = "obrien_fleming"
    futility_boundary: float = 0.10


@dataclass
class BioequivalenceDesign:
    """
    Bioequivalence study design.

    Attributes:
        n_periods: Number of periods (default: 2)
        n_sequences: Number of sequences (default: 2)
        washout_duration: Washout period in days
        bioequivalence_limits: BE acceptance limits
        parameters: PK parameters for BE assessment
        regulatory_guidance: Regulatory guidance ('fda' or 'ema')

    Example:
        >>> design = BioequivalenceDesign(regulatory_guidance='fda')
    """
    n_periods: int = 2
    n_sequences: int = 2
    washout_duration: float = 7.0
    bioequivalence_limits: Tuple[float, float] = (0.80, 1.25)
    parameters: List[str] = field(default_factory=lambda: ["cmax", "auc_0_inf"])
    regulatory_guidance: str = "fda"


def parallel_design(
    n_arms: int,
    randomization_ratio: Optional[List[float]] = None,
    stratification_factors: Optional[List[str]] = None,
) -> ParallelDesign:
    """
    Create a parallel group study design.

    Args:
        n_arms: Number of treatment arms
        randomization_ratio: Randomization ratio (default: equal)
        stratification_factors: Stratification factors

    Returns:
        ParallelDesign object

    Example:
        >>> design = parallel_design(2)
        >>> design = parallel_design(3, randomization_ratio=[0.5, 0.25, 0.25])
    """
    return ParallelDesign(
        n_arms=n_arms,
        randomization_ratio=randomization_ratio or [],
        stratification_factors=stratification_factors or [],
    )


def crossover_2x2(washout_duration: float = 7.0) -> CrossoverDesign:
    """
    Create a 2-period, 2-sequence crossover design (AB, BA).

    Args:
        washout_duration: Washout period in days

    Returns:
        CrossoverDesign object

    Example:
        >>> design = crossover_2x2(washout_duration=14.0)
    """
    return CrossoverDesign(
        n_periods=2,
        n_sequences=2,
        washout_duration=washout_duration,
        sequence_assignments=[[1, 2], [2, 1]],
    )


def crossover_3x3(washout_duration: float = 7.0) -> CrossoverDesign:
    """
    Create a 3-period, 3-sequence crossover design (Latin square).

    Args:
        washout_duration: Washout period in days

    Returns:
        CrossoverDesign object

    Example:
        >>> design = crossover_3x3(washout_duration=21.0)
    """
    return CrossoverDesign(
        n_periods=3,
        n_sequences=3,
        washout_duration=washout_duration,
        sequence_assignments=[[1, 2, 3], [2, 3, 1], [3, 1, 2]],
    )


def williams_design(n_treatments: int, washout_duration: float = 7.0) -> CrossoverDesign:
    """
    Create a Williams design for crossover studies.

    Williams designs are balanced for first-order carryover effects.

    Args:
        n_treatments: Number of treatments (2, 3, or 4)
        washout_duration: Washout period in days

    Returns:
        CrossoverDesign object

    Example:
        >>> design = williams_design(4, washout_duration=14.0)
    """
    if n_treatments == 2:
        sequences = [[1, 2], [2, 1]]
    elif n_treatments == 3:
        sequences = [[1, 2, 3], [2, 3, 1], [3, 1, 2],
                     [1, 3, 2], [3, 2, 1], [2, 1, 3]]
    elif n_treatments == 4:
        sequences = [[1, 2, 4, 3], [2, 3, 1, 4], [3, 4, 2, 1], [4, 1, 3, 2],
                     [1, 4, 2, 3], [2, 1, 3, 4], [3, 2, 4, 1], [4, 3, 1, 2]]
    else:
        raise ValueError("Williams design only supports 2, 3, or 4 treatments")

    return CrossoverDesign(
        n_periods=n_treatments,
        n_sequences=len(sequences),
        washout_duration=washout_duration,
        sequence_assignments=sequences,
    )


def dose_escalation_3plus3(
    dose_levels: List[float],
    starting_dose: Optional[float] = None,
    max_dlt_rate: float = 0.33,
    cohort_size: int = 3,
    max_subjects: int = 30,
) -> DoseEscalationDesign:
    """
    Create a 3+3 dose escalation design.

    Args:
        dose_levels: Available dose levels
        starting_dose: Starting dose (default: first level)
        max_dlt_rate: Maximum acceptable DLT rate
        cohort_size: Cohort size (default: 3)
        max_subjects: Maximum total subjects

    Returns:
        DoseEscalationDesign object

    Example:
        >>> design = dose_escalation_3plus3([10.0, 25.0, 50.0, 100.0, 200.0])
    """
    return DoseEscalationDesign(
        dose_levels=dose_levels,
        starting_dose=starting_dose,
        escalation_rule="3+3",
        cohort_size=cohort_size,
        max_subjects=max_subjects,
        target_dlt_rate=max_dlt_rate,
    )


def dose_escalation_mtpi(
    dose_levels: List[float],
    starting_dose: Optional[float] = None,
    target_dlt_rate: float = 0.25,
    cohort_size: int = 3,
    max_subjects: int = 30,
) -> DoseEscalationDesign:
    """
    Create an mTPI dose escalation design.

    Args:
        dose_levels: Available dose levels
        starting_dose: Starting dose (default: first level)
        target_dlt_rate: Target DLT rate
        cohort_size: Cohort size
        max_subjects: Maximum total subjects

    Returns:
        DoseEscalationDesign object

    Example:
        >>> design = dose_escalation_mtpi([10.0, 25.0, 50.0, 100.0], target_dlt_rate=0.30)
    """
    return DoseEscalationDesign(
        dose_levels=dose_levels,
        starting_dose=starting_dose,
        escalation_rule="mTPI",
        cohort_size=cohort_size,
        max_subjects=max_subjects,
        target_dlt_rate=target_dlt_rate,
    )


def dose_escalation_crm(
    dose_levels: List[float],
    starting_dose: Optional[float] = None,
    target_dlt_rate: float = 0.25,
    cohort_size: int = 1,
    max_subjects: int = 30,
) -> DoseEscalationDesign:
    """
    Create a CRM (Continual Reassessment Method) dose escalation design.

    Args:
        dose_levels: Available dose levels
        starting_dose: Starting dose (default: first level)
        target_dlt_rate: Target DLT rate
        cohort_size: Cohort size (usually 1 for CRM)
        max_subjects: Maximum total subjects

    Returns:
        DoseEscalationDesign object

    Example:
        >>> design = dose_escalation_crm([10.0, 25.0, 50.0, 100.0, 200.0])
    """
    return DoseEscalationDesign(
        dose_levels=dose_levels,
        starting_dose=starting_dose,
        escalation_rule="CRM",
        cohort_size=cohort_size,
        max_subjects=max_subjects,
        target_dlt_rate=target_dlt_rate,
    )


def adaptive_design(
    base_design: Union[ParallelDesign, CrossoverDesign],
    interim_analyses: List[float] = [0.5],
    alpha_spending: str = "obrien_fleming",
    futility_boundary: float = 0.10,
) -> AdaptiveDesign:
    """
    Create an adaptive trial design.

    Args:
        base_design: Underlying study design
        interim_analyses: Information fractions for interim analyses
        alpha_spending: Alpha spending function ('obrien_fleming', 'pocock', 'haybittle_peto')
        futility_boundary: Futility boundary (conditional power threshold)

    Returns:
        AdaptiveDesign object

    Example:
        >>> base = parallel_design(2)
        >>> design = adaptive_design(base, interim_analyses=[0.5])
    """
    return AdaptiveDesign(
        base_design=base_design,
        interim_analyses=interim_analyses,
        alpha_spending=alpha_spending,
        futility_boundary=futility_boundary,
    )


def bioequivalence_design(
    n_periods: int = 2,
    n_sequences: int = 2,
    washout_duration: float = 7.0,
    bioequivalence_limits: Tuple[float, float] = (0.80, 1.25),
    parameters: List[str] = ["cmax", "auc_0_inf"],
    regulatory_guidance: str = "fda",
) -> BioequivalenceDesign:
    """
    Create a bioequivalence study design.

    Args:
        n_periods: Number of periods
        n_sequences: Number of sequences
        washout_duration: Washout period in days
        bioequivalence_limits: BE acceptance limits
        parameters: PK parameters to assess
        regulatory_guidance: Regulatory guidance ('fda' or 'ema')

    Returns:
        BioequivalenceDesign object

    Example:
        >>> design = bioequivalence_design(regulatory_guidance='fda')
        >>> # For highly variable drugs (EMA)
        >>> design = bioequivalence_design(
        ...     bioequivalence_limits=(0.6984, 1.4319),
        ...     regulatory_guidance='ema'
        ... )
    """
    return BioequivalenceDesign(
        n_periods=n_periods,
        n_sequences=n_sequences,
        washout_duration=washout_duration,
        bioequivalence_limits=bioequivalence_limits,
        parameters=parameters,
        regulatory_guidance=regulatory_guidance,
    )


def get_design_description(design: Any) -> str:
    """
    Get a human-readable description of a study design.

    Args:
        design: Study design object

    Returns:
        Design description string

    Example:
        >>> design = parallel_design(2)
        >>> print(get_design_description(design))
        '2-arm parallel group design'
    """
    if isinstance(design, ParallelDesign):
        return f"{design.n_arms}-arm parallel group design"
    elif isinstance(design, CrossoverDesign):
        return f"{design.n_periods}-period {design.n_sequences}-sequence crossover design"
    elif isinstance(design, DoseEscalationDesign):
        return f"Dose escalation ({design.escalation_rule}) with {len(design.dose_levels)} dose levels"
    elif isinstance(design, AdaptiveDesign):
        base_desc = get_design_description(design.base_design)
        n_interim = len(design.interim_analyses)
        return f"Adaptive {base_desc} with {n_interim} interim analysis(es)"
    elif isinstance(design, BioequivalenceDesign):
        return f"Bioequivalence {design.n_periods}x{design.n_sequences} crossover ({design.regulatory_guidance})"
    else:
        return str(type(design).__name__)
